node/block_exchange.go

```diff
@@ -0,0 +1,226 @@
+package node
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"github.com/celestiaorg/go-header"
+	goheaderp2p "github.com/celestiaorg/go-header/p2p"
+	goheaderstore "github.com/celestiaorg/go-header/store"
+	goheadersync "github.com/celestiaorg/go-header/sync"
+	ds "github.com/ipfs/go-datastore"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/net/conngater"
+	"github.com/tendermint/tendermint/libs/log"
+	tmtypes "github.com/tendermint/tendermint/types"
+	"go.uber.org/multierr"
+
+	"github.com/rollkit/rollkit/config"
+	"github.com/rollkit/rollkit/p2p"
+	"github.com/rollkit/rollkit/types"
+)
+
+type BlockExchangeService struct {
+	conf         config.NodeConfig
+	genesis      *tmtypes.GenesisDoc
+	p2p          *p2p.Client
+	ex           *goheaderp2p.Exchange[*types.Block]
+	syncer       *goheadersync.Syncer[*types.Block]
+	sub          *goheaderp2p.Subscriber[*types.Block]
+	p2pServer    *goheaderp2p.ExchangeServer[*types.Block]
+	blockStore   *goheaderstore.Store[*types.Block]
+	syncerStatus *SyncerStatus
+
+	logger log.Logger
+	ctx    context.Context
+}
+
+func NewBlockExchangeService(ctx context.Context, store ds.TxnDatastore, conf config.NodeConfig, genesis *tmtypes.GenesisDoc, p2p *p2p.Client, logger log.Logger) (*HeaderExchangeService, error) {
+	// store is TxnDatastore, but we require Batching, hence the type assertion
+	// note, the badger datastore impl that is used in the background implements both
+	storeBatch, ok := store.(ds.Batching)
+	if !ok {
+		return nil, errors.New("failed to access the datastore")
+	}
+	ss, err := goheaderstore.NewStore[*types.SignedHeader](storeBatch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the header store: %w", err)
+	}
+
+	return &HeaderExchangeService{
+		conf:         conf,
+		genesis:      genesis,
+		p2p:          p2p,
+		ctx:          ctx,
+		headerStore:  ss,
+		logger:       logger,
+		syncerStatus: new(SyncerStatus),
+	}, nil
+}
+
+func (bExService *BlockExchangeService) initBlockStoreAndStartSyncer(ctx context.Context, initial *types.Block) error {
+	if err := bExService.blockStore.Init(ctx, initial); err != nil {
+		return err
+	}
+	if err := bExService.syncer.Start(bExService.ctx); err != nil {
+		return err
+	}
+	bExService.syncerStatus.m.Lock()
+	defer bExService.syncerStatus.m.Unlock()
+	bExService.syncerStatus.started = true
+	return nil
+}
+
+func (bExService *BlockExchangeService) tryInitBlockStoreAndStartSyncer(ctx context.Context, trustedBlock *types.Block) {
+	if trustedBlock != nil {
+		if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
+			bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
+		}
+	}
```

Comment:
nit: I think always trying to keep code more inline is more readable.

```suggestion
	if trustedBlock == nil {
                return
	}
	if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
		bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
	}
```

Additionally, why do we not return the error and let the caller decide how to handle it? 
I think if we are handling errors within a function and not returning anything, we should have a clear docstring explaining the logic of the function and how errors are handled. 
Or we should defer the logic to the caller and just return an error.

Reply:
Talked about this with Ganesh, this method was used as a goroutine before which is why he didn't return errors and only logged them when writing HeaderExchange. We decided that this method can be blocking so modified it to return errors as well.